src/rpc/rawtransaction.cpp

```diff
@@ -575,44 +578,78 @@ static RPCHelpMan decodescript()
     }
     ScriptPubKeyToUniv(script, r, /* include_hex */ false);
 
-    UniValue type;
-    type = find_value(r, "type");
-
-    if (type.isStr() && type.get_str() != "scripthash") {
-        // P2SH cannot be wrapped in a P2SH. If this script is already a P2SH,
-        // don't return the address for a P2SH of the P2SH.
-        r.pushKV("p2sh", EncodeDestination(ScriptHash(script)));
-        // P2SH and witness programs cannot be wrapped in P2WSH, if this script
-        // is a witness program, don't return addresses for a segwit programs.
-        if (type.get_str() == "pubkey" || type.get_str() == "pubkeyhash" || type.get_str() == "multisig" || type.get_str() == "nonstandard") {
-            std::vector<std::vector<unsigned char>> solutions_data;
-            TxoutType which_type = Solver(script, solutions_data);
-            // Uncompressed pubkeys cannot be used with segwit checksigs.
-            // If the script contains an uncompressed pubkey, skip encoding of a segwit program.
-            if ((which_type == TxoutType::PUBKEY) || (which_type == TxoutType::MULTISIG)) {
-                for (const auto& solution : solutions_data) {
-                    if ((solution.size() != 1) && !CPubKey(solution).IsCompressed()) {
-                        return r;
-                    }
-                }
-            }
-            UniValue sr(UniValue::VOBJ);
-            CScript segwitScr;
-            if (which_type == TxoutType::PUBKEY) {
-                segwitScr = GetScriptForDestination(WitnessV0KeyHash(Hash160(solutions_data[0])));
-            } else if (which_type == TxoutType::PUBKEYHASH) {
-                segwitScr = GetScriptForDestination(WitnessV0KeyHash(uint160{solutions_data[0]}));
-            } else {
-                // Scripts that are not fit for P2WPKH are encoded as P2WSH.
-                // Newer segwit program versions should be considered when then become available.
-                segwitScr = GetScriptForDestination(WitnessV0ScriptHash(script));
+    std::vector<std::vector<unsigned char>> solutions_data;
+    const TxoutType which_type{Solver(script, solutions_data)};
+
+    // Check if wrapping makes sense at all
+    switch (which_type) {
+    case TxoutType::MULTISIG:
+    case TxoutType::NONSTANDARD:
+    case TxoutType::PUBKEY:
+    case TxoutType::PUBKEYHASH:
+    case TxoutType::WITNESS_V0_KEYHASH:
+    case TxoutType::WITNESS_V0_SCRIPTHASH:
+        // Can be wrapped
+        break;
+    case TxoutType::NULL_DATA:
+    case TxoutType::SCRIPTHASH:
+    case TxoutType::WITNESS_UNKNOWN:
+    case TxoutType::WITNESS_V1_TAPROOT:
+        // Should not be wrapped, so return early
+        return r;
```

Comment:
Not sure this "return early" is a good approach. Seems like it would be better to have a conditional block.